packages/compartment-mapper/src/policy.js

```diff
@@ -0,0 +1,305 @@
+        `Importing '${specifier}' was not allowed by policy packages:${JSON.stringify(
+          policy.packages,
+        )}`,
+      );
+    }
+    return;
+  }
+
+  if (!policy.builtins || !policy.builtins[specifier]) {
+    throw Error(
+      `Importing '${specifier}' was not allowed by policy 'builtins':${JSON.stringify(
+        policy.builtins,
+      )}`,
+    );
+  }
+};
+
+function attenuateModule({ attenuators, name, params, originalModule }) {
+  if (!attenuators) {
+    // TODO: figure out what to pass here or where to move this check
+    throw Error(
+      `Attenuation '${name}' in policy doesn't have a corresponding implementation.`,
+    );
+  }
+  const attenuationCompartment = new Compartment(
+    {},
+    {},
+    {
+      name,
+      resolveHook: moduleSpecifier => moduleSpecifier,
+      importHook: async () => {
+        const {
+          namespace: { attenuate },
+        } = await attenuators(name);
+        const ns = await attenuate(params, originalModule);
+        const staticModuleRecord = freeze({
+          imports: [],
+          exports: keys(ns),
+          execute: moduleExports => {
+            assign(moduleExports, ns);
+          },
+        });
+        return staticModuleRecord;
+      },
+    },
+  );
+  return attenuationCompartment.module('.');
+}
+
+/**
+ * Throws if importing of the specifier is not allowed by the policy
+ *
+ * @param {string} specifier - exit module name
+ * @param {Object} originalModule - reference to the exit module
+ * @param {Object} policy - local compartment policy
+ * @param {Object} attenuators - a key-value where attenuations can be found
+ */
+export const attenuateModuleHook = (
+  specifier,
+  originalModule,
+  policy,
+  attenuators,
+) => {
+  if (policy && (!policy.builtins || !policy.builtins[specifier])) {
+    throw Error(
+      `Attenuation failed '${specifier}' was not in policy builtins:${stringify(
+        policy.builtins,
+      )}`,
+    );
+  }
+  if (!policy || policy.builtins[specifier] === true) {
+    return originalModule;
+  }
+
+  return attenuateModule({
+    attenuators,
+    name: policy.builtins[specifier].attenuate,
+    params: policy.builtins[specifier].params,
+    originalModule,
+  });
+};
+
+const padDiagnosis = text => ` (${text})`;
+/**
+ * Provide dignostic information for a missing compartment error
+ *
+ * @param {Object}  params
+ * @param {string}  params.moduleSpecifier
+ * @param {Object}  params.compartmentDescriptor
+ * @param {string}  params.foreignModuleSpecifier
+ * @param {string}  params.foreignCompartmentName
+ * @returns {string}
+ */
+export const diagnoseMissingCompartmentError = ({
+  moduleSpecifier,
+  compartmentDescriptor,
+  foreignModuleSpecifier,
+  foreignCompartmentName,
+}) => {
+  const { policy, name, scopes } = compartmentDescriptor;
+
+  if (policy) {
+    if (!policy.packages) {
+      return padDiagnosis(
+        `There were no allowed packages specified in policy for "${name}"`,
```

Comment:
Please use a `q` function for robust quoting of values here and below.  It can be as simple as `const q = JSON.stringify;` if the normal `const { quote: q } = assert;` convention doesn't work.

```suggestion
        `There were no allowed packages specified in policy for ${q(name)}`,
```